### PR TITLE
Adjust model to allow multiple polygons per title [WIP]

### PR DESCRIPTION
**Pull request by davidread**

### CODE REVIEW

File: landavailability/lr/management/commands/import_polygons.py [hunk 1]
```diff
@@ -1,31 +1,70 @@
...
+        self.title_ids = set()
+        ShapefileImportCommand(*kargs, **kwargs)
+
+    def process_record(self, polygon_id, title_id, insert_date, update_date,
+                       status, geometry):
+
+        def create_polygon():
+            poly = Polygon()
+            self.polygon_ids.add(polygon_id)
+            poly.id = polygon_id
```
Comment (andreagrandi) [inline landavailability/lr/management/commands/import_polygons.py]: lines 24 and 26 could be `poly = Polygon(id=polygon_id)`

File: landavailability/lr/management/commands/import_polygons.py [hunk 2]
```diff
@@ -1,31 +1,70 @@
...
+            poly = create_polygon()
             outcome = 'created'
-            poly.title = record.record[1]
 
-        poly.insert = datetime.strptime(record.record[2], '%Y-%m-%dT%H:%M:%S')
-        poly.update = datetime.strptime(record.record[3], '%Y-%m-%dT%H:%M:%S')
-        poly.status = record.record[4]
+        def create_title():
+            title = Title()
+            title.id = title_id
```
Comment (andreagrandi) [inline landavailability/lr/management/commands/import_polygons.py]: same here: `title = Title(id=title_id)`

File: landavailability/lr/serializers.py [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
 
 
-class LRPolySerializer(serializers.ModelSerializer):
+class TitleSerializer(serializers.ModelSerializer):
 
     class Meta:
-        model = LRPoly
-        fields = ('title', 'insert', 'update', 'status', 'geom')
+        model = Polygon
+        fields = ('id')
```
Comment (andreagrandi) [inline landavailability/lr/serializers.py:9]: Reminder for me to fix this in the API ticket

File: landavailability/tests/lr/test_models.py [hunk 4]
```diff
@@ -1,15 +1,18 @@
...
 from datetime import datetime
 
 
 class TestLandModel(TestCase):
     @pytest.mark.django_db
     def test_lrpoly_model_creation(self):
-        lrpoly = LRPoly()
-        lrpoly.title = 'A6523948'
+        lrpoly = Polygon()
+        lrpoly.id = 1234
```
Comment (andreagrandi) [inline landavailability/tests/lr/test_models.py]: `lrpoly = Polygon(id=1234)`

File: landavailability/tests/lr/test_models.py [hunk 5]
```diff
@@ -39,12 +42,15 @@ def test_lrpoly_model_creation(self):
...
-        self.assertEqual(LRPoly.objects.count(), 1)
+        self.assertEqual(Polygon.objects.count(), 1)
 
     @pytest.mark.django_db
     def test_uprn_model_creation(self):
-        lrpoly = LRPoly()
-        lrpoly.title = 'A6523948'
+        lrpoly = Polygon()
+        lrpoly.id = 1234
+        title = Title(id='A6523948')
```
Comment (andreagrandi) [inline landavailability/tests/lr/test_models.py]: `lrpoly = Polygon(id=1234)`

**Review by andreagrandi — CHANGES_REQUESTED** [review]
Minor changes commented but not really necessary. Up to you otherwise I can merge as is.

**Review by andreagrandi — APPROVED** [review]

**Comment by davidread** [comment]
I did some stats, and the optimization (caching polygon and title ids in memory, avoiding some db look-ups) speeds up the import by 31%.

```
Without optimization
{'created': 10001} Count: 10001 Rate: 685000/hour
{'created': 50001} Count: 50001 Rate: 694000/hour
{'created': 100001} Count: 100001 Rate: 696000/hour
{'created': 200001} Count: 200001 Rate: 702000/hour

With optimization
{'created': 10001} Count: 10001 Rate: 777000/hour
{'created': 50001} Count: 50001 Rate: 868000/hour
{'created': 100001} Count: 100001 Rate: 897000/hour
{'created': 200001} Count: 200001 Rate: 922000/hour
```

**Pull request merged**